tools/navcon-converter/navcon-converter

```diff
@@ -0,0 +1,147 @@
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+# ===========================================================================
+
+import struct
+import sys
+
+"""
+Version 2 binary format
+
+Header version number then connection count,
+then arrays of various types.
+The userids field is useless.
+
+Example:
+header count {x1 z1 y1 x2 z2 y2}[] rad[] flags[] areas[] dirs[] userids[]
+i      i      f  f  f  f  f  f     f     H       B       B      I
+
+int header;
+int count;
+struct coords_t { float x1, z1, y1, x2, z2, y2 };
+coords_t[count] coords;
+float rad[count];
+unsigned short flags[count];
+unsigned byte areas[count];
+unsigned byte dirs[count];
+unsigned int userids[count];
+"""
+
+"""
+Version 3 text format
+
+Header line with prefix and integer version number,
+then one connection per line, all numbers as integers.
+There is no useless userids field.
+
+Example:
+navcon 3
+x1 z1 y1 x2 z2 y2 rad flags areas dirs
+[…]
+"""
+
+header_text_prefix = "navcon"
+current_format_version = 3
+
+def read_unpack(file_blob, value_format):
+    value_size = struct.calcsize(value_format)
+
+    if isinstance(file_blob, str):
+        value_bytes = bytes(file_blob[:value_size], "utf-8")
+    else:
+        value_bytes = file_blob.read(value_size)
+
+    return int(struct.unpack(value_format, value_bytes)[0])
+
+def read_format2(file_blob):
+    format_dict = {
+        "x1": "f",
+        "z1": "f",
+        "y1": "f",
+        "x2": "f",
+        "z2": "f",
+        "y2": "f",
+        "rad": "f",
+        "flags": "H",
+        "areas": "B",
+        "dirs": "B",
+        "userids": "I",
+    }
+
+    coords_list = ["x1", "z1", "y1", "x2", "z2", "y2"]
+
+    print("Converting {}".format(file_name), file=sys.stderr)
+    print("Converting from format version {} to {}".format(header_magic, current_format_version), file=sys.stderr)
+
+    connection_count = read_unpack(file_blob, "<i")
+    print("Converting {} connections".format(str(connection_count)), file=sys.stderr)
+
+    connection_list = [{} for i in range(0, connection_count)]
+
+    for key_name in format_dict.keys():
+        for i in range(0, connection_count):
+            if key_name == coords_list[0]:
+                subkey_list = coords_list
+            elif key_name in coords_list[1:]:
+                continue
+            else:
+                subkey_list = [key_name]
+
+            for subkey_name in subkey_list:
+                value_int = read_unpack(file_blob, format_dict[subkey_name])
+
+                if subkey_name != "userids":
+                    connection_list[i][subkey_name] = value_int
+
+    file_blob.close()
+
+    return connection_list
+
+def write_format3(connection_list):
+    print("{} {}".format(header_text_prefix, str(current_format_version)))
+
+    for connection in connection_list:
+        print(*connection.values())
```

Comment:
Actually the `%.0f` thing (see the comment on the C++ side) would be good here too for correct rounding. You can do `print("%.0f %.0f %.0f etc" % tuple(connection.values())`

Reply:
I done it in another way, calling `round()` before converting to `int` (they're already `int` at this point).